pkg/clients/nodepool/nodepool.go

```diff
@@ -0,0 +1,331 @@
+
+		if in.Config.SandboxConfig != nil && spec.Config.SandboxConfig == nil {
+			spec.Config.SandboxConfig = &v1beta1.SandboxConfig{
+				SandboxType: in.Config.SandboxConfig.SandboxType,
+			}
+		}
+
+		spec.Config.ServiceAccount = gcp.LateInitializeString(spec.Config.ServiceAccount, in.Config.ServiceAccount)
+
+		if in.Config.ShieldedInstanceConfig != nil {
+			if spec.Config.ShieldedInstanceConfig == nil {
+				spec.Config.ShieldedInstanceConfig = &v1beta1.ShieldedInstanceConfig{}
+			}
+			spec.Config.ShieldedInstanceConfig.EnableIntegrityMonitoring = gcp.LateInitializeBool(spec.Config.ShieldedInstanceConfig.EnableIntegrityMonitoring, in.Config.ShieldedInstanceConfig.EnableIntegrityMonitoring)
+			spec.Config.ShieldedInstanceConfig.EnableSecureBoot = gcp.LateInitializeBool(spec.Config.ShieldedInstanceConfig.EnableSecureBoot, in.Config.ShieldedInstanceConfig.EnableSecureBoot)
+		}
+
+		spec.Config.Tags = gcp.LateInitializeStringSlice(spec.Config.Tags, in.Config.Tags)
+
+		if len(in.Config.Taints) != 0 && len(spec.Config.Taints) == 0 {
+			spec.Config.Taints = make([]*v1beta1.NodeTaint, len(in.Config.Taints))
+			for i, t := range in.Config.Taints {
+				spec.Config.Taints[i] = &v1beta1.NodeTaint{
+					Effect: t.Effect,
+					Key:    t.Key,
+					Value:  t.Value,
+				}
+			}
+		}
+
+		if in.Config.WorkloadMetadataConfig != nil && spec.Config.WorkloadMetadataConfig == nil {
+			spec.Config.WorkloadMetadataConfig = &v1beta1.WorkloadMetadataConfig{
+				NodeMetadata: in.Config.WorkloadMetadataConfig.NodeMetadata,
+			}
+		}
+	}
+
+	spec.InitialNodeCount = gcp.LateInitializeInt64(spec.InitialNodeCount, in.InitialNodeCount)
+	spec.Locations = gcp.LateInitializeStringSlice(spec.Locations, in.Locations)
+
+	if in.Management != nil {
+		if spec.Management == nil {
+			spec.Management = &v1beta1.NodeManagementSpec{}
+		}
+
+		spec.Management.AutoRepair = gcp.LateInitializeBool(spec.Management.AutoRepair, in.Management.AutoRepair)
+		spec.Management.AutoUpgrade = gcp.LateInitializeBool(spec.Management.AutoUpgrade, in.Management.AutoUpgrade)
+	}
+
+	if in.MaxPodsConstraint != nil && spec.MaxPodsConstraint == nil {
+		spec.MaxPodsConstraint = &v1beta1.MaxPodsConstraint{
+			MaxPodsPerNode: in.MaxPodsConstraint.MaxPodsPerNode,
+		}
+	}
+
+	spec.Version = gcp.LateInitializeString(spec.Version, in.Version)
+}
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(in *v1beta1.NodePoolParameters, currentState container.NodePool) (bool, UpdateKind) {
+	currentParams := &v1beta1.NodePoolParameters{}
+	LateInitializeSpec(currentParams, currentState)
+	if !cmp.Equal(in.Autoscaling, currentParams.Autoscaling) {
+		return false, AutoscalingUpdate
+	}
+	if !cmp.Equal(in.Management, currentParams.Management) {
+		return false, ManagementUpdate
+	}
+	// if !cmp.Equal(in.InitialNodeCount, currentParams.InitialNodeCount) {
```

Comment:
We can't actually compare `InitialNodeCount` to get the size of the cluster. To get the size, we need to loop through `InstanceGroupURLs` and see how many nodes each has, by calling the compute API for instance groups.

Reply:
I think we shouldn't worry about `InitialNodeCount` field at all since it's only useful for the first `Create` call. Our `spec` represents the desired state and after the creation `Initial*` is not a part of the _desired_ state. So, I'd not account for that field for updates.

(This is yet another example that we need validation for immutable fields.)

Reply:
Cool! You'll see in the most recent commit that the comparison of the current spec and the late initialized spec (used to determine if we need a "general update") ignores `InitialNodeCount`